backend/cn/solver.mli

```diff
@@ -1,25 +1,36 @@
+(* Module Solver -- Interface to the SMT solver via SMTLIB *)
 type solver
 type model
+(* (TODO: BCP: The "with quantifiers" part will be the instantiations that the
+   solver found -- is that right?) *)
 type model_with_q = model * (Sym.t * LogicalSorts.t) list
-
 val empty_model : model
+
+(* Global flags to pass to the solver (TODO: BCP: Could use a bit more
+   documentation, maybe) *)
 val random_seed : int ref
 val log_to_temp : bool ref
 val log_dir     : (string option) ref
 val solver_path : (string option) ref
 val solver_flags : ((string list) option) ref
 val solver_type : (Simple_smt.solver_extensions option) ref
 
+(* Create a solver *)
 val make : Global.t -> solver
 
-val num_scopes : solver -> int
-val push : solver -> unit
-val pop : solver -> int -> unit
+(* Incrementally (and imperatively) add an assumption to the solver state *)
 val add_assumption : solver -> Global.t -> LogicalConstraints.t -> unit
 
-val set_slow_smt_settings : float option -> string option -> unit
+(* Save / restore solver state, to support backtracking *)
+val push : solver -> unit
+val pop : solver -> int -> unit
 
+(* TODO: BCP: What is this? *)
+val num_scopes : solver -> int
 
+(* Run the solver.  Note that we pass the assumptions explicitly even
+   though they are also available in the solver context, because CN is
+   going some simplification on its own. *)
```

Comment:
Plus we have a default quantifier instantiation heuristic, which adds additional constraints derived from `assumptions` based on the to-be-proved constraint.